src/shared/mesh/MeshBuilder.cpp

```diff
@@ -0,0 +1,162 @@
+#include "MeshBuilder.hpp"
+
+#include "axom/slic.hpp"
+#include "axom/primal.hpp"
+
+namespace shared {
+
+MeshBuilder MeshBuilder::Merged( std::initializer_list<MeshBuilder> meshes )
+{
+  std::vector<mfem::Mesh*> mesh_list;
+  mesh_list.reserve( meshes.size() );
+  for ( auto& mesh : meshes ) {
+    // NOTE: the const cast is because the constructor requires a non-const mesh, even though the data is copied and not
+    // altered
+    mesh_list.push_back( &const_cast<mfem::Mesh&>( *mesh ) );
+  }
+  return mfem::Mesh( mesh_list.data(), mesh_list.size() );
+}
+
+MeshBuilder MeshBuilder::SquareMesh( int n_x_els, int n_y_els )
+{
+  return mfem::Mesh::MakeCartesian2D( n_x_els, n_y_els, mfem::Element::QUADRILATERAL );
+}
+
+MeshBuilder MeshBuilder::CubeMesh( int n_x_els, int n_y_els, int n_z_els )
+{
+  return mfem::Mesh::MakeCartesian3D( n_x_els, n_y_els, n_z_els, mfem::Element::HEXAHEDRON );
+}
+
+MeshBuilder MeshBuilder::HypercubeMesh( int dim, int n_els )
+{
+  switch ( dim ) {
+    case 1:
+      return mfem::Mesh::MakeCartesian1D( n_els );
+    case 2:
+      return mfem::Mesh::MakeCartesian2D( n_els, n_els, mfem::Element::QUADRILATERAL );
+    case 3:
+      return mfem::Mesh::MakeCartesian3D( n_els, n_els, n_els, mfem::Element::HEXAHEDRON );
+    default:
+      SLIC_ERROR_ROOT( "Mesh dimension must be between 1 and 3." );
+      return mfem::Mesh();
+  }
+}
+
+MeshBuilder::MeshBuilder( mfem::Mesh&& mesh ) : mesh_{ std::move( mesh ) } { mesh_.EnsureNodes(); }
+
+MeshBuilder&& MeshBuilder::translate( std::initializer_list<double> dx )
+{
+  SLIC_ERROR_ROOT_IF( static_cast<int>( dx.size() ) != mesh_.SpaceDimension(), "Invalid size for dx" );
+  auto& coords = *mesh_.GetNodes();
+  for ( int d = 0; d < mesh_.SpaceDimension(); ++d ) {
+    for ( int i = 0; i < mesh_.GetNV(); ++i ) {
+      auto vdof = coords.FESpace()->DofToVDof( i, d );
+      coords[vdof] += *( dx.begin() + d );
+    }
+  }
+  return std::move( *this );
+}
+
+MeshBuilder&& MeshBuilder::updateAttrib( int old_attrib, int new_attrib )
+{
+  for ( int i = 0; i < mesh_.GetNE(); ++i ) {
+    if ( mesh_.GetAttribute( i ) == old_attrib ) {
+      mesh_.SetAttribute( i, new_attrib );
+    }
+  }
+  auto old_attrib_idx = mesh_.attributes.Find( old_attrib );
```

Comment:
It sort of looks like these lines of code outside the loop over number of elements should do what `SetAttribute()` above does. Maybe not a big deal. A comment here might clarify what is being done.

Reply:
I'll add a comment